### PR TITLE
GCI-2179 Implement remove basket item endpoint

### PR DESCRIPTION
**Pull request by cgregoryCH**
* Adds an endpoint to the basket controller to remove items within a basket.

[GCI-2179](https://companieshouse.atlassian.net/browse/GCI-2179)

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 1]
```diff
@@ -80,7 +79,10 @@ public class BasketController {
             "${uk.gov.companieshouse.orders.api.basket.checkouts}";
     public static final String PATCH_PAYMENT_DETAILS_URI =
             "${uk.gov.companieshouse.orders.api.basket.checkouts}/{id}/payment";
-
+    public static final String REMOVE_ITEM_URI =
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:82]: Please remove this comment

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 2]
```diff
@@ -434,6 +436,53 @@ public ResponseEntity<String> patchBasketPaymentDetails(final @RequestBody Baske
         return ResponseEntity.status(NO_CONTENT).body(null);
     }
 
+    @PostMapping(REMOVE_ITEM_URI) // Change to PUT?
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:439]: I would use the PUT request method for this endpoint.

**Review by kinpang-CH — APPROVED** [review]

File: src/test/java/uk/gov/companieshouse/orders/api/interceptor/UserAuthorisationInterceptorTests.java [hunk 3]
```diff
@@ -339,6 +339,26 @@ void preHandleAcceptsAuthorisedPostReprocessOrderRequest() {
         thenRequestIsAccepted();
     }
 
+    @Test
+    @DisplayName("preHandle accepts get basket links request that has ")
```
Comment (dlgroves-ch) [inline src/test/java/uk/gov/companieshouse/orders/api/interceptor/UserAuthorisationInterceptorTests.java:343]: Incomplete test name

File: src/test/java/uk/gov/companieshouse/orders/api/interceptor/UserAuthorisationInterceptorTests.java [hunk 4]
```diff
@@ -339,6 +339,26 @@ void preHandleAcceptsAuthorisedPostReprocessOrderRequest() {
...
+    void getBasketLinks() {
+        // given
+        givenRequest(GET, "/basket/links");
+
+        // when and then
+        thenRequestIsAccepted();
+    }
+
+    @Test
+    @DisplayName("preHandle accepts get basket links request that has ")
```
Comment (dlgroves-ch) [inline src/test/java/uk/gov/companieshouse/orders/api/interceptor/UserAuthorisationInterceptorTests.java:353]: Incomplete test name

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 5]
```diff
@@ -434,6 +437,53 @@ public ResponseEntity<String> patchBasketPaymentDetails(final @RequestBody Baske
...
+        String itemUri = basketRequestDTO.getItemUri();
+        logMap.put(LoggingUtils.ITEM_URI, itemUri);
+
+        String id = EricHeaderHelper.getIdentity(request);
+        final Optional<Basket> retrievedBasket = basketService.getBasketById(id);
+
+        if (retrievedBasket.isPresent()) {
+            if (!basketService.removeBasketDataItemByUri(id, basketRequestDTO.getItemUri())) {
+                LOGGER.error("Item not found for item_uri: " + basketRequestDTO.getItemUri(), new ResourceNotFoundException("Item uri not found"), logMap);
+                return ResponseEntity.status(NOT_FOUND).body(null); // failure
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:456]: Can be simplified:
```java
return ResponseEntity.notFound().build();
```

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 6]
```diff
@@ -434,6 +437,53 @@ public ResponseEntity<String> patchBasketPaymentDetails(final @RequestBody Baske
...
+            }
+            logMap.put(LoggingUtils.BASKET_ID, id);
+            logMap.put(LoggingUtils.STATUS, HttpStatus.OK);
+            LOGGER.infoRequest(request, "Removed item from basket", logMap);
+            return ResponseEntity.status(HttpStatus.OK).body(null); // success
```
Comment (dlgroves-ch) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:461]: Can be simplified:
```java
return ResponseEntity.ok().build();
```

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]
Several return values in BasketController can be simplified, e.g. HTTP 404 with no response body can be simplified as `ResponseEntity.notFound().build()`.

**Review by sthompsonCH — APPROVED** [review]

**Review by dlgroves-ch — APPROVED** [review]

**Pull request merged**